### PR TITLE
adguardhome: add support for platforms other than x86_64-linux

### LINKED ISSUE #113900: pkgs.adguardhome has wrong platform attribute
**Describe the bug**

AdGuardHome has the meta platform `platforms.linux`, but is a [precompiled ELF binary for x86_64](https://github.com/NixOS/nixpkgs/blob/037936b7a307c7399cf0f3d9fabe37ea5b0b8534/pkgs/servers/adguardhome/default.nix#L8).

**To Reproduce**
Try to run `adguardhome` on `aarch64` (or any architecture other than `x86_64`). It will fail with `Exec format error`:

```
# nix-shell --pure -p adguardhome

[nix-shell:~]# adguardhome
bash: /nix/store/c08rij11vr8jcp5v2qcngr0mggzx34q9-adguardhome/bin/adguardhome: cannot execute binary file: Exec format error
```

**Expected behavior**

`adguardhome` should execute, so that it can be used.

**Notify maintainers**

@numkem 

**Metadata**

 - system: `"aarch64-linux"`
 - host os: `Linux 5.4.92, NixOS, 21.03pre265961.891f607d530 (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.10`
 - channels(root): `"nixos-21.03pre265961.891f607d530"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
 - adguardhome
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by iagocq**
###### Motivation for this change

To fix #113900 and make adguardhome run on more than just x86_64-linux.

###### Things done

This downloads the correct binary from the AdGuardHome Github page depending on the host platform. An update.sh script was also added to update the derivation to newer versions with little effort, and also to add more host systems that have a binary release upstream but currently aren't listed here.

I tested the resulting binary on my desktop (x86_64-linux) and a Raspberry Pi 3B (aarch64-linux), both running NixOS. Both ran the binary through the nixos/adguardhome module fine.

While building from source would be ideal, there are many problems in doing that because of how the upstream project is organized (npm, yarn (frontend) and go (backend)). I could only figure out how to build the frontend correctly, but that required not using the makefile provided by the project.

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/servers/adguardhome/default.nix [hunk 1]
```diff
@@ -16,8 +13,8 @@ stdenv.mkDerivation rec {
   meta = with lib; {
     homepage = "https://github.com/AdguardTeam/AdGuardHome";
     description = "Network-wide ads & trackers blocking DNS server";
-    platforms = platforms.linux;
-    maintainers = with maintainers; [ numkem ];
+    platforms = platforms.all;
+    maintainers = with maintainers; [ numkem iagoq ];
     license = licenses.gpl3;
```
Comment (SuperSandro2000) [inline pkgs/servers/adguardhome/default.nix:18]: gpl3Plus or gpl3Only?

File: pkgs/servers/adguardhome/version.nix [hunk 2]
```diff
@@ -0,0 +1 @@
+"0.106.3"
```
Comment (SuperSandro2000) [inline pkgs/servers/adguardhome/version.nix:1]: Can we remove this extra file and sed the version in the default.nix?

File: pkgs/servers/adguardhome/update.sh [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+systems[linux_amd64]=x86_64-linux
+systems[linux_arm64]=aarch64-linux
+systems[darwin_amd64]=x86_64-darwin
+
+echo 'fetchurl:' > "$bins"
+echo '{'        >> "$bins"
+
+for asset in $(curl --silent https://api.github.com/repos/AdguardTeam/AdGuardHome/releases/latest | jq -c '.assets[]') ; do
+    url="$(jq -r '.browser_download_url' <<< "$asset")"
+    adg_system="$(egrep -o '(darwin|linux)_(386|amd64|arm64)' <<< "$url" || echo -n)"
```
Comment (SuperSandro2000) [inline pkgs/servers/adguardhome/update.sh:28]:
```suggestion
    adg_system="$(grep -Eo '(darwin|linux)_(386|amd64|arm64)' <<< "$url" || echo -n)"
```

File: pkgs/servers/adguardhome/update.sh [hunk 4]
```diff
@@ -0,0 +1,37 @@
+#! /usr/bin/env nix-shell
+#! nix-shell -i bash -p nix-prefetch jq
```
Comment (SuperSandro2000) [inline pkgs/servers/adguardhome/update.sh:2]:
```suggestion
#! nix-shell -i bash -p curl gnugrep nix-prefetch jq
```

File: pkgs/servers/adguardhome/default.nix [hunk 5]
```diff
@@ -16,8 +13,8 @@ stdenv.mkDerivation rec {
   meta = with lib; {
     homepage = "https://github.com/AdguardTeam/AdGuardHome";
     description = "Network-wide ads & trackers blocking DNS server";
-    platforms = platforms.linux;
-    maintainers = with maintainers; [ numkem ];
+    platforms = platforms.all;
```
Comment (SuperSandro2000) [inline pkgs/servers/adguardhome/default.nix:16]:
```suggestion
    platforms = platforms.unix;
```

File: pkgs/servers/adguardhome/default.nix (hunk 1 above)
Comment (iagocq) [reply pkgs/servers/adguardhome/default.nix:18]: It's gpl3Only apparently. No mentions of later versions in their repository.

File: pkgs/servers/adguardhome/version.nix (hunk 2 above)
Comment (iagocq) [reply pkgs/servers/adguardhome/version.nix:1]: Would be nice, will get to it.

File: maintainers/maintainer-list.nix [hunk 6]
```diff
@@ -4169,6 +4169,16 @@
     githubId = 2789926;
     name = "Imran Hossain";
   };
+  iagoq = {
+    email = "18238046+iagocq@users.noreply.github.com";
+    github = "iagocq";
+    githubId = 18238046;
+    name = "Iago Manoel Brito";
+    keys = [{
+      longkattributeattributeattributeeyid = "rsa4096/0x35D39F9A9A1BC8DA";
```
Comment (iagocq) [inline maintainers/maintainer-list.nix:4178]: I have no idea how this happened

File: pkgs/servers/adguardhome/default.nix [hunk 7]
```diff
@@ -1,13 +1,10 @@
-{ lib, stdenv, fetchurl }:
+{ lib, stdenv, fetchurl, system ? stdenv.targetPlatform }:
 
 stdenv.mkDerivation rec {
-  name = "adguardhome";
-  version = "0.105.2";
+  pname = "adguardhome";
+  version = "v0.106.3";
```
Comment (SuperSandro2000) [inline pkgs/servers/adguardhome/default.nix:5]:
```suggestion
  version = "0.106.3";
```

version numbers in nix shouldn't start with a v.

Comment (iagocq) [reply pkgs/servers/adguardhome/default.nix:5]: Oh I missed that when committing, sorry

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 124026` at d11d65f7 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>adguardhome</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/servers/adguardhome/default.nix:9:3:
```
  |
9 |   installPhase = ''
  |   ^
```
</li>
  </ul>
</details>

**Review by numkem — APPROVED** [review]
lgtm

**Comment by Artturin** [comment]
Please use rebase fixup to combine some of the commits https://git-rebase.io/#fixup

And get rid of the merge commit, to update your fork without having a merge commit see https://medium.com/@topspinj/how-to-git-rebase-into-a-forked-repo-c9f05e821c8a

**Comment by iagocq** [comment]
I forgot about this PR, lol.
I think this is better now.

**Comment by Artturin** [comment]
eval check fails on aarch64-darwin
the solution is to add https://github.com/AdguardTeam/AdGuardHome/releases/download/v0.107.0-b.11/AdGuardHome_darwin_arm64.zip

File: pkgs/servers/adguardhome/default.nix [hunk 8]
```diff
@@ -1,23 +1,24 @@
...
+    updateScript = ./update.sh;
+  };
+
   meta = with lib; {
     homepage = "https://github.com/AdguardTeam/AdGuardHome";
     description = "Network-wide ads & trackers blocking DNS server";
-    platforms = platforms.linux;
-    maintainers = with maintainers; [ numkem ];
-    license = licenses.gpl3;
+    platforms = platforms.unix;
```
Comment (Artturin) [inline pkgs/servers/adguardhome/default.nix:20]:
```suggestion
    platforms = [ "x86_64-linux" "aarch64-linux" "x86_64-darwin" "aarch64-darwin" ];
```

Comment (iagocq) [reply pkgs/servers/adguardhome/default.nix]: sadly there isn’t a binary package for aarch64-darwin in the latest release version. I changed `platforms` to only include the ones we download the bins for though

Comment (Artturin) [reply pkgs/servers/adguardhome/default.nix]: okay

**Comment by Artturin** [comment]
fixup the latest commit

**Comment by Artturin** [comment]
Thanks

**Comment by iagocq** [comment]
thank you!

**Pull request merged**